### ISSUE TITLE
Support for trusted proxy headers

### ISSUE THREAD

**Issue by harlan-zw** [unjs/h3#504]
### Describe the feature

As a module author, for maximum compatibility, it makes life easier to just use the proxy headers and fallback to node specifics. The issue is that these won't be safe in all environments and can be spoofed (i.e `x-forwarded-for`)

To solve this we need a simple implementation that:
1. allows users to provide a list of trusted proxies
2. allows module authors to generate the headers using this list of trusted proxies

I think combining 1 and 2 may be difficult but maybe you have some ideas on that. As a simple first implementation a `getRequestTrustedHeaders(event, trustedProxies)` may be enough.

Inspo: https://laravel.com/docs/master/requests#configuring-trusted-proxies

### Additional information

- [X] Would you be willing to help implement this feature?

**Comment by pi0** [comment]
We have `getRequestHost({ xForwardedHost: true })`. Main benefit of using utils for each semi-standard header likt this is that we can support other vendor header as same name. How do you feel about introducing smaller utils like `getRequestIP` instead?

**Comment by harlan-zw** [comment]
Not sure if this explanation is answering the issue about trusting easily spoofed headers. 

H3 should provide a small util function to only get the trusted values, similar to https://github.com/symfony/http-foundation/blob/6.3/Request.php#L783

It would then be on frameworks to allow configuring the trusted proxies and on modules to use that config.

**Comment by pi0** [comment]
I am up to supporting a special event context property such as `event.context.isTrustedProxy` to automatically enable headers detection however it should be user responsibility solely to write logic. Based on server setup it is different how to determine trustibility.

**Comment by harlan-zw** [comment]
> I am up to supporting a special event context property such as `event.context.isTrustedProxy` to automatically enable headers detection however it should be user responsibility solely to write logic. Based on server setup it is different how to determine trustibility.

Yes, I'm not concerned about users' usage, but about any integration code relying on these headers to work.

Consider this as an example: Nitro implements an in-built rate limiter, to effectively rate limit we need to know the trusted IP of the request. Nitro has no current way to know if it can trust the connecting IP, generically. This is because the x-forwarded-for header can easily be spoofed on some servers without proxies that use it.

Now blow this example up to Nuxt modules and all the other ways we need to know trusted IPs then you should have a better understanding of the problem. The solution requires multiple layers to solve, but as a starting point, we could introduce a way to filter for trusted headers similar to https://github.com/symfony/http-foundation/blob/6.3/Request.php#L2004.

**Comment by Atinux** [comment]
I really like this discussion :)

I like the `event.context.trusedHeaders` option, not sure if the `trustedHeaders` could be given when creating the app though? https://github.com/unjs/h3/blob/588c8a3d9b1e408f36a1f51a43d759534245029f/src/app.ts#L50-L62

I don't know in term of performance if this will have an impact or not

**Comment by harlan-zw** [comment]
> @harlan-zw Would that make sense if `event.context.trustedProxies: string[]` was introduced at the h3 level, and if the array is not empty, when using `getHeaders` or anything headers related, h3 would check if the ip is in the trustedProxies ?
> 
> Then it would be Nuxt/Nitro/Modules responsibility to set `trustedProxies` (in a plugin or in a middleware)
> 
> Doing that way might make it easier for adoption.

Seems like a good way to do it :+1:

**Comment by acidjazz** [comment]
what ever came of this?

**Comment by pi0** [comment]
`trustProxy` is being added to srvx (https://github.com/h3js/srvx/pull/223) so every consumer can consistently trust on `req.url`
